### PR TITLE
[PyT] Plumbing correct bias dims from TE to cudnn, while adding support for additional bias shapes

### PR DESCRIPTION
**Pull request by KshitijLakhani**
# Description

TE common was not plumbing attention vector bias dimensions correctly to cuDNN.
Instead of using shape from Bias, i.e. `[bias_sq, bias_skv]` it was using `[sq, skv]` thereby passing larger than required dims. This PR correctly plumbs the bias shape from TE PyT to cuDNN via TE common.

Additionally, this PR also adds support for `dbias` , i.e, bias grad (fwd+bwd) calculation for b1ss, bhss, 11ss (initially only 1hss was supported) for CP and non-CP cases.
Support for `bias` calculation , i.e. no bias grad (fwd only) for 111s is also added for CP and non-CP cases
(bwd support to be added once cuDNN start supporting it in the future - TODOs sprinkled in code for the same)

Lastly, tests are added to support all newly added functionality for both CP and non-CP cases

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Infra/Build change
- [ ] Code refactoring

## Changes

- Pass `bias_sq` and `bias_skv` to `fused_attn_arbitrary_seqlen_fwd_impl()` and `fused_attn_arbitrary_seqlen_bwd_impl()` 
- Add new entries for `bias_sq` and `bias_skv` in `FADescriptor_v1`
- Correct the bias passed to the MHA cuDNN graph to use `bias_sq` and `bias_skv` instead of `s_q` and `s_kv` 
- Enable dbias calculation for all cuDNN supported shapes : 1hss, 11ss, b1ss, bhss
- Add TODOs for when cuDNN starts supporting dbias calculation for bias shape 111s

## Testing:
- Added tests (fwd only and no bias grad) for 111s bias shape in both, non-CP and CP fused attn tests
- Added tests for 1hss, b1ss, bhss, 111s bias shapes in CP fused attn tests (non-CP already has tests for all other supported shapes)
- Confirmed by using `NVTE_DEBUG` and additional test logging that the same test bias shape passes from PyT layer to cuDNN (this was necessary as there were hard coded shapes that would show a false positive thereby masking actual behavior) 

## Supplementary testing: 

Using the reproducer : https://github.com/cyanguwa/TransformerEngine/tree/test_111s for bias [1,1,1,s] it can be seen in the cuDNN FE logs that prior to this PR the bias dims passed onto cuDNN from TE were
`{"data_type":null,"dim":[1,1,128,128],"is_pass_by_value":false,"is_virtual":false,"name":"bias","pass_by_value":null,"reordering_type":"NONE","stride":[16384,16384,128,1],"uid":0,"uid_assigned":false},` 
and after this PR they are: 
`"bias":{"data_type":null,"dim":[1,1,1,128],"is_pass_by_value":false,"is_virtual":false,"name":"bias","pass_by_value":null,"reordering_type":"NONE","stride":[128,128,128,1],"uid":0,"uid_assigned":false},`

# Checklist:

- [X] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [X] The functionality is complete
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by cyanguwa** [comment]
Looks good - please pick the 111s test from my branch as well. Thanks!

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu [hunk 1]
```diff
@@ -1224,11 +1237,15 @@ void fused_attn_arbitrary_seqlen_bwd(
...
   size_t bias_h = 0;
+  size_t bias_sq = 0;
+  size_t bias_skv = 0;
   if ((bias_type != NVTE_Bias_Type::NVTE_NO_BIAS) && (bias_type != NVTE_Bias_Type::NVTE_ALIBI)) {
     devPtrBias = input_Bias->data.dptr;
     devPtrdBias = output_dBias->data.dptr;
     bias_b = output_dBias->data.shape[0];
     bias_h = output_dBias->data.shape[1];
+    bias_sq = input_Bias->data.shape[2];
+    bias_skv = input_Bias->data.shape[3];
```
[the remark this answers is not in the document]
Comment (KshitijLakhani) [reply transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu]: Addressed in SHA 143ede5cc34b3c33b2b70da4422096f6d41478c1

File: tests/pytorch/attention/run_attention_with_cp.py [hunk 2]
```diff
@@ -467,6 +469,22 @@ def run_dpa_with_cp(
...
         ]
+        if dbias is not None and dbias_ is not None:
+            dbias = dbias.view(
+                dbias.shape[0],
+                dbias.shape[1],
+                2 * world_size,
+                dbias.shape[2] // (2 * world_size),
+                dbias.shape[3],
+            )
+            # bias has fixed axis (2) as dbias shape: (1, 1, max_seqlen_q, max_seqlen_kv)
```
Comment (cyanguwa) [inline tests/pytorch/attention/run_attention_with_cp.py]: I think our CP implementation (after your C changes) should support all bias shapes, not just 111s. I also _think_ your reshaping here should work for all shapes. Could you run the tests to confirm?

File: tests/pytorch/attention/run_attention_with_cp.py [hunk 3]
```diff
@@ -433,23 +435,23 @@ def run_dpa_with_cp(
             out_.backward(dout_fp8_)
         else:
             out_.backward(dout_)
-    dq_, dk_, dv_ = q_.grad, k_.grad, v_.grad
+    dq_, dk_, dv_, dbias_ = q_.grad, k_.grad, v_.grad, bias_.grad
```
[the remark this answers is not in the document]
Comment (KshitijLakhani) [reply tests/pytorch/attention/run_attention_with_cp.py]: Fixed

File: tests/pytorch/attention/run_attention_with_cp.py [hunk 4]
```diff
@@ -338,7 +339,7 @@ def run_dpa_with_cp(
             out.backward(dout_fp8)
         else:
             out.backward(dout)
-    dq, dk, dv = q.grad, k.grad, v.grad
+    dq, dk, dv, dbias = q.grad, k.grad, v.grad, bias.grad
```
[the remark this answers is not in the document]
Comment (KshitijLakhani) [reply tests/pytorch/attention/run_attention_with_cp.py]: Fixed

File: tests/pytorch/attention/test_attention_with_cp.py [hunk 5]
```diff
@@ -257,6 +286,10 @@ def test_cp_with_fused_attention(
         pytest.skip("FP8 attention cannot work with sliding window yet!")
     if "p2p" in cp_comm_type and config.window_size != (-1, 0) and config.window_size != (-1, -1):
         pytest.skip("CP implementation with KV P2P does not support sliding window yet!")
+    if "p2p" in cp_comm_type and config.attn_bias_type != "no_bias" and config.bias_shape == "111s":
+        pytest.skip(
+            f"CP implementation with KV P2P requires bias sequence dim to be divisible by 2"
```
Comment (cyanguwa) [inline tests/pytorch/attention/test_attention_with_cp.py]: I feel this skip message doesn't make sense. Should it be because we don't support dbias for 111s? But we can skip the bwd for it right?

File: transformer_engine/pytorch/attention/dot_product_attention/utils.py [hunk 6]
```diff
@@ -958,12 +958,11 @@ def _is_fa3_supported(num_heads, num_gqa_groups, head_dim_qk, head_dim_v, qkv_dt
...
-            # [1, 1, s, s], [b, 1, s, s] and [b, h, s, s]
-            logger.debug("Disabling FusedAttention for dBias in [1, H, S, S] shape")
+        # dbias calculation is not supported for 111s as of cuDNN 9.18. So, use fused attention backend only if bias does not require grad.
+        if fu_core_attention_bias_requires_grad and fu_core_attention_bias_shape == "111s":
+            logger.warning("Disabling FusedAttention as dbias calculation is not supported for 111s")
             use_fused_attention = False
-        else:
+        if not fu_core_attention_bias_requires_grad:
             # max512 backend will only support [1, h, s, s]
             os.environ["NVTE_FUSED_ATTN_BACKEND"] = "1"
```
[the remark this answers is not in the document]
Comment (cyanguwa) [reply transformer_engine/pytorch/attention/dot_product_attention/utils.py]: I think it should be "elif" here. Could you double check the logic here please?

**Comment by cyanguwa** [comment]
Please address the couple of questions I left, but otherwise, it looks good! Thanks!

File: tests/pytorch/attention/test_attention_with_cp.py (hunk 5 above)
Comment (KshitijLakhani) [reply tests/pytorch/attention/test_attention_with_cp.py]: You are right about the skipping of bwd for the non-CP part.
For non-CP, this PR skips the testing of bwd and only test fwd for bias 111s. This works completely fine.
This is achieved by setting `is_training=False` and `core_attention_bias_requires_grad=False`

However, if the same is done for CP, i.e. set the two above shown flags to false, the test asserts at: https://github.com/NVIDIA/TransformerEngine/blob/3774aa37e2aa777c0f3bf3dca752d64f961725e1/transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py#L1450 , which requires the bias seq dim to be divisible by 2, which will never happen right as bias is 111s.
So my understanding is that the only CP comm type, p2p, which supports bias, cannot actually support it (even in the fwd pass) due to the requirements in : https://github.com/NVIDIA/TransformerEngine/blob/3774aa37e2aa777c0f3bf3dca752d64f961725e1/transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py#L1450 

So, that's why I just skip that test using the same assert message as a skip message.

I can make the skip message more descriptive my directly attributing this divisibility requirement not being fulfilled to the bias 111s shape by saying something as below: 
`f"CP implementation with KV P2P requires bias sequence dim to be divisible by 2, which is not possible with {config.bias_shape=}"`

The reason I suggest we use the above skip message is :
i) The skip is due to the fundamental reason that the seq dim for bias is not divisible by 2
ii) To your point, which I agree with, point i) stems as a consequence of the bias shape being 111s and not because the bias dim is just explicitly set to a number not divisible by 2.

Let me know your thoughts @cyanguwa

Comment (cyanguwa) [reply tests/pytorch/attention/test_attention_with_cp.py]: Ok. That makes sense. Could you make the message a bit more descriptive please, like you suggested? For 111s + CP, I wonder if we should be broadcasting the `sq` dimension instead of slicing it?

File: transformer_engine/pytorch/attention/dot_product_attention/utils.py [hunk 7]
```diff
@@ -966,12 +966,13 @@ def _is_fa3_supported(num_heads, num_gqa_groups, head_dim_qk, head_dim_v, qkv_dt
...
+        # dbias calculation is not supported for 111s as of cuDNN 9.18. So, use fused attention backend only if bias does not require grad.
+        if fu_core_attention_bias_requires_grad and fu_core_attention_bias_shape == "111s":
+            logger.warning(
+                "Disabling FusedAttention as dbias calculation is not supported for 111s"
+            )
             use_fused_attention = False
-        else:
+        if not fu_core_attention_bias_requires_grad:
             # max512 backend will only support [1, h, s, s]
             os.environ["NVTE_FUSED_ATTN_BACKEND"] = "1"
```
[the remark this answers is not in the document]
Comment (KshitijLakhani) [reply transformer_engine/pytorch/attention/dot_product_attention/utils.py:978]: Addressed

File: tests/pytorch/attention/test_attention_with_cp.py (hunk 5 above)
Comment (KshitijLakhani) [reply tests/pytorch/attention/test_attention_with_cp.py]: Removed this skip
Added support for 111s bias in CP fused attn
Added support in CP fused attn tests to set is_training to False and no grad

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py [hunk 8]
```diff
@@ -1442,20 +1453,30 @@ def forward(
...
-                *attn_bias.shape[:-2],
-                2,
-                attn_bias.shape[-2] // 2,
-                2 * cp_size,
-                attn_bias.shape[-1] // (2 * cp_size),
+                "Only support bias shape of [1,1,sq,skv], [1,h,sq,skv], [b,1,sq,skv], [b,h,sq,skv], [1,1,sq,skv] for forward, "
+                "and [1,1,sq,skv], [1,h,sq,skv], [b,1,sq,skv], [b,h,sq,skv] for backward!"
             )
+            # For all bias shapes except 111s, sq must be divisible by 2 and sk must be divisible by 2*cp_size
+            # For bias shape 111s, only sq must be divisible by 2
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: Should it be:
For bias shape 111s, only skv must be divisible by 2*cp_size

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py [hunk 9]
```diff
@@ -1442,20 +1453,30 @@ def forward(
...
+                attn_bias_ = attn_bias.view(
+                    *attn_bias.shape[:-2],
+                    2,
+                    attn_bias.shape[-2] // 2,
+                    2 * cp_size,
+                    attn_bias.shape[-1] // (2 * cp_size),
+                )
+            else:
+                assert attn_bias.shape[-1] % (2 * cp_size) == 0, "Sequence length does not meet divisible requirements!"
+                # [b, h, sq, sk] -> [b, h, sq, 2*cp, sk//(2*cp)]
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py:1478]: Will you be missing a dimension, or will it be fine if no index_select is done?

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py (hunk 8 above)
Comment (KshitijLakhani) [reply transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: Thanks for catching this 
Fixed it

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py (hunk 9 above)
Comment (KshitijLakhani) [reply transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py:1478]: If I understand right, your concern is if the consumer of `attn_bias_` is aware of the missing dimensions or not ? Let me try answering this:

So, the only consumer of this `attn_bias_` is in `cp_p2p_fwd_fused_attn()` 

In `cp_p2p_fwd_fused_attn()` lines L844-860 take care of the concatenation based on 111s shape vs non-111s shape for bias. After the concatenation, the final shape created for `attn_bias_inputs` is [b, h, sq_chunk, sk_chunk], with sq_chunk being 1 for 111s, and, non-1 for non-111s shapes.
And I believe, this is fine as the broadcast for this is then handled downstream.

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu [hunk 10]
```diff
@@ -812,19 +818,20 @@ void fused_attn_arbitrary_seqlen_bwd_impl(
...
-        // shapes [1, 1, s, s], [b, 1, s, s], [b, h, s, s]
-        // are not supported for dbias calculation but they are
-        // supported for forward bias calculation
-        if ((bias_b == 1) && (bias_h == h)) {
+        // bias shapes [1, 1, s, s], [b, 1, s, s], [b, h, s, s], [1, h, s, s] are supported for dbias calculation
+        // bias shape [1, 1, 1, s] is not supported for dbias calculation as of cuDNN 9.18
+        if (!((bias_b == 1) && (bias_h == 1) && (bias_sq == 1))) {
           sdpa_backward_options.set_dbias(dBias);
         }
       }
```
[the remark this answers is not in the document]
Comment (KshitijLakhani) [reply transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu]: Unrelated to this PR, however, I think it is a good change to create the tensor as needed.
Added a change for this

File: tests/pytorch/attention/run_attention_with_cp.py [hunk 11]
```diff
@@ -428,138 +472,261 @@ def run_dpa_with_cp(
...
+                    # Compare dbias (same as BSHD)
+                    if names[i] == "dbias":
+                        # After reshaping: (1, 1, 2, seq_q//2, seq_kv)
+                        # Compare along dimension 2 (the split sequence dimension)
+                        ndim_bias = t.ndim
+                        seq_q_dim_bias = ndim_bias - 2
+                        slice_0 = [slice(None)] * ndim_bias
+                        slice_0[seq_q_dim_bias] = 0
+                        slice_1 = [slice(None)] * ndim_bias
+                        slice_1[seq_q_dim_bias] = 1
```
[the remark this answers is not in the document]
Comment (KshitijLakhani) [reply tests/pytorch/attention/run_attention_with_cp.py:682]: Good catch by greptile
Was comparing the dim 3/-2 of the 5d tensor instead of dim 2/-3 of the tensor
Re ran the test and all pass locally

File: tests/pytorch/attention/run_attention_with_cp.py [hunk 12]
```diff
@@ -428,138 +472,261 @@ def run_dpa_with_cp(
...
+                        # After reshaping: (1, 1, 2, seq_q//2, seq_kv)
+                        # Compare along dimension 2 (the split sequence dimension)
+                        ndim_bias = t.ndim
+                        seq_q_dim_bias = ndim_bias - 2  # Query sequence dimension
+                        # After reshaping both have shape: [..., 2, seq_q//2, seq_kv]
+                        # The split dimension is at seq_q_dim_bias
+                        slice_0 = [slice(None)] * ndim_bias
+                        slice_0[seq_q_dim_bias] = 0
+                        slice_1 = [slice(None)] * ndim_bias
+                        slice_1[seq_q_dim_bias] = 1
```
[the remark this answers is not in the document]
Comment (KshitijLakhani) [reply tests/pytorch/attention/run_attention_with_cp.py:628]: Good catch by greptile
Was comparing the dim 3/-2 of the 5d tensor instead of dim 2/-3 of the tensor
Re ran the test and all pass locally

File: tests/pytorch/attention/test_attention_with_cp.py [hunk 13]
```diff
@@ -147,7 +147,10 @@ def test_cp_with_flash_attention(dtype, model, qkv_format, cp_comm_type):
         2, 4096, 12, 128, attn_mask_type="causal", attn_bias_type="post_scale_bias"
     ),  # MHA
     "cp_1_3": ModelConfig(2, 4096, 12, 128, attn_bias_type="post_scale_bias"),  # MHA
-    "cp_1_4": ModelConfig(2, 4096, 12, 128, attn_mask_type="causal", window_size=(512, 512)),  # MHA
+    "cp_1_4": ModelConfig(
+        2, 4096, 12, 128, attn_bias_type="post_scale_bias", bias_shape="bhss"
+    ),  # MHA
+    "cp_1_5": ModelConfig(2, 4096, 12, 128, attn_mask_type="causal", window_size=(512, 0)),  # MHA
```
[the remark this answers is not in the document]
Comment (KshitijLakhani) [reply tests/pytorch/attention/test_attention_with_cp.py]: This is a good catch by greptile as well 
Seems like it was altered during testing and never reinstated :( 
Done in [ff174a8](https://github.com/NVIDIA/TransformerEngine/pull/2537/commits/ff174a8a080e392f334ee8fce7b769370fdd352c)
